gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/public/src/components/selectionTree/js/SelectionTree.js

```diff
@@ -0,0 +1,442 @@
+            var segs = gpii.oauth2.selectionTree.expandSegs(requestedPref);
+            var newVal = fluid.get(newModel, segs);
+            var oldVal = fluid.get(oldModel, segs);
+
+            if (newVal !== oldVal) {
+                var elm = that.container.find(that.options.domMap[requestedPref]);
+                that.updateDOMState(elm, newVal);
+            }
+        });
+    };
+
+    gpii.oauth2.selectionTree.DOMSetup = function (that) {
+        fluid.each(that.options.domMap, function (selector, selectorName) {
+            var elm = that.container.find(selector);
+            if (that.options.requestedPrefs[selectorName] || selectorName === "") {
+                // bind change event to update model (checked/unchecked)
+                that.bindDOMListener(elm, selectorName);
+            } else {
+                that.removeLeaf(elm);
+            }
+        });
+
+        // bind modelListener to update dom (checked/indeterminate)
+        that.applier.modelChanged.addListener("", that.updateDOMFromModel);
+    };
+
+    gpii.oauth2.selectionTree.updateModel = function (that, ELPath, state) {
+        var model = fluid.copy(that.model);
+
+        gpii.oauth2.selectionTree.setAllDescendants(fluid.get(model, ELPath), state);
+        gpii.oauth2.selectionTree.setAncestors(model, ELPath);
+
+        that.applier.change("", model);
+    };
+
+    gpii.oauth2.selectionTree.bindDOMListener = function (that, elm, ELPath, updateModelFn) {
+        elm.change(function (e) {
+            var state = $(e.target).prop("checked") ? "checked" : "unchecked";
+            updateModelFn(ELPath, state);
+        });
+    };
+
+    gpii.oauth2.selectionTree.relayClick = function (elm) {
+        $(elm).click();
+    };
+
+    gpii.oauth2.selectionTree.toggleClass = function (elm, classes, state) {
+        $(elm).toggleClass(classes, state);
+    };
+
+    gpii.oauth2.selectionTree.removeLeaf = function (elm, leafSel) {
+        elm.closest(leafSel).remove();
+    };
+
+    gpii.oauth2.selectionTree.setCheckbox = function (checkbox, state) {
+        var checked = state === "checked";
+        var indeterminate = state === "indeterminate";
+
+        checkbox.prop({
+            "indeterminate": indeterminate,
+            "checked": checked
+        });
+    };
+
+    gpii.oauth2.selectionTree.gatherPaths = function (model) {
+        var pathValue = model.value;
+        var paths = [];
+
+        if (pathValue === "checked") {
+            paths.push("");
+        } else if (pathValue === "indeterminate") {
+            fluid.each(model, function (subModel, seg) {
+                if (fluid.isPlainObject(subModel)) {
+                    var subPaths = gpii.oauth2.selectionTree.gatherPaths(subModel);
+                    fluid.each(subPaths, function (path) {
+                        if (path) {
+                            paths.push(seg + "." + path);
+                        } else {
+                            paths.push(seg);
```

Comment:
This method of terminating the recursion is pretty irregular. Factorise this into two functions, a top-level driver which accepts model, dereferences value and sets up path, and a separate function capturing the recursion. The recursion termination should occur at an outer level of this function (hopefully simply as a byproduct of the type check - clearer to use negation of isPrimitive rather than isPlainObject) rather than so deeply nested